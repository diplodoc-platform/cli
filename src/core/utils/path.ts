import {dirname, isAbsolute, join, normalize} from 'node:path';
import _normalizePath from 'normalize-path';

import {isExternalHref} from '~/core/utils/url';

export function normalizePath(path: string): NormalizedPath {
    return _normalizePath(normalize(_normalizePath(path, false)), false);
}

export function isRelativePath(path: string): path is RelativePath {
    return !isExternalHref(path) && !isAbsolute(path);
}

export function langFromPath(path: string, config: {lang?: string; langs: string[]}) {
    const {lang, langs} = config;
    const pathBaseLang = normalizePath(path).split('/')[0];
    const pathLang = langs.includes(pathBaseLang) && pathBaseLang;

    return pathLang || lang || langs[0];
}

export function rebasePath(root: RelativePath, path: RelativePath) {
    return normalizePath(join(dirname(root), path));
}

export function fullPath(
    path: AbsolutePath | NormalizedPath,
    root: NormalizedPath,
): NormalizedPath {
    if (path.match(/^(\/|\\)/)) {
        return normalizePath(path.slice(1));
    } else {
        return normalizePath(join(dirname(root), path));
    }
}
