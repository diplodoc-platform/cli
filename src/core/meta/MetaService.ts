import type {Run} from '~/core/run';
import type {Meta, Resources} from './types';

import {omit, uniq} from 'lodash';

import {copyJson, normalizePath} from '~/core/utils';

import {getHooks, withHooks} from './hooks';

type Config = {
    addSystemMeta: boolean;
};

@withHooks
export class MetaService {
    readonly name = 'Meta';

    // private run: Run<Config>;

    private config: Config;

    private meta: Map<NormalizedPath, Meta> = new Map();

    constructor(run: Run<Config>) {
        // this.run = run;
        this.config = run.config;
    }

    /**
     * Returns non normalized current readonly metadata for selected path.
     */
    get(path: RelativePath) {
        const file = normalizePath(path);
        return copyJson(this.meta.get(file)) || this.initialMeta();
    }

    /**
     * Returns normalized merged metadata for selected path.
     */
    async dump(path: RelativePath) {
        const file = normalizePath(path);
        const meta = copyJson(this.meta.get(file)) || this.initialMeta();

        for (const field of ['script', 'style', 'keywords'] as const) {
            if (!meta[field]) {
                continue;
            }

            meta[field] = uniq(meta[field] as string[]).filter(Boolean);
        }

        for (const field of ['script', 'style', 'keywords', 'contributors', 'csp'] as const) {
            if (!meta[field]?.length) {
                delete meta[field];
            }
        }

        for (const field of ['metadata', '__system']) {
            if (!meta[field]) {
                continue;
            }

            if (!Object.keys(meta[field] as Hash).length) {
                delete meta[field];
            }
        }

        return getHooks(this).Dump.promise(meta, file);
    }

    add(path: RelativePath, record: Hash) {
        const file = normalizePath(path);

        const meta = this.meta.get(file) || this.initialMeta();
        const result = Object.assign(
            meta,
            omit(record, ['script', 'style', 'csp', 'metadata', '__system']),
        );

        this.meta.set(file, result);

        this.addMetadata(path, record.metadata);
        this.addSystemVars(path, record.__system);

        return meta;
    }

    addResources(path: RelativePath, resources: Resources | undefined) {
        const file = normalizePath(path);

        if (!resources) {
            return;
        }

        const meta = this.meta.get(file) || this.initialMeta();

        if (Array.isArray(resources.script)) {
            meta.script = uniq([...(meta.script || []), ...resources.script]);
        }

        if (Array.isArray(resources.style)) {
            meta.style = uniq([...(meta.style || []), ...resources.style]);
        }

        if (Array.isArray(resources.csp)) {
            for (const record of resources.csp) {
                for (const [key, value] of Object.entries(record)) {
                    record[key] = ([] as string[]).concat(value);
                }
            }

            meta.csp = [...(meta.csp || []), ...resources.csp];
        }

        this.meta.set(file, meta);
    }

    addMetadata(path: RelativePath, metadata: Hash | undefined) {
        const file = normalizePath(path);

        if (!metadata) {
            return;
        }

        if (!Array.isArray(metadata)) {
            metadata = Object.entries(metadata).map(([name, content]) => ({name, content}));
        }

        const meta = this.meta.get(file) || this.initialMeta();
        meta.metadata = metadata.concat(meta.metadata || []);

        this.meta.set(file, meta);
    }

    addSystemVars(path: RelativePath, vars: Hash | undefined) {
        const file = normalizePath(path);

        if (!vars || !this.config.addSystemMeta) {
            return;
        }

        const meta = this.meta.get(file) || this.initialMeta();
        meta.__system = Object.assign({}, meta.__system, vars);

        this.meta.set(file, meta);
    }

    private initialMeta(): Meta {
        return {
            metadata: [],
            style: [],
            script: [],
            csp: [],
        };
    }
}
