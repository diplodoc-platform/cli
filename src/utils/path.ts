import {sep} from 'path';
import {Platforms} from '../constants';

export function addSlashPrefix(path: string): string {
    const slashPrefix = path.startsWith(sep) ? '' : sep;

    return `${slashPrefix}${path}`;
}

export function convertBackSlashToSlash(path: string): string {
    if (process.platform === Platforms.WINDOWS) {
        return path.replace(/\\/g, '/');
    }

    return path;
}
