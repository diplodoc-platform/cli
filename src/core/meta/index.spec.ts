import {join} from 'node:path';
import {describe, expect, it} from 'vitest';
import {when} from 'vitest-when';
import {dedent} from 'ts-dedent';

import {setupRun} from '~/commands/build/__tests__';

import {MetaService} from './MetaService';
import {uniq} from 'lodash';

const content = {
    'test.md': dedent`
        ---
        metadata:
        - name: generator
          content: Diplodoc Platform vDIPLODOC-VERSION
        ---
        test text
    `,
};

function prepare() {
    const run = setupRun({
        addSystemMeta: false,
    });

    const service = new MetaService(run);

    for (const [file, data] of Object.entries(content)) {
        when(run.read).calledWith(join(run.input, file)).thenResolve(data);
    }

    return service;
}

async function call(path: RelativePath) {
    const service = prepare();

    expect(service.dump(path)).toMatchSnapshot();
}

function test(name: string, path: string) {
    it(name, async () => call(path as RelativePath));
}

describe('meta', () => {
    describe('service', () => {
        describe('load', () => {
            test('should load test file', 'test.md');

            it('should handle `has` trap', async () => {
                const service = prepare();

                service.addMetadata('test.md' as RelativePath, {
                    generator: `Diplodoc Platform vDIPLODOC-VERSION`,
                });
                service.addMetadata('test.md' as RelativePath, {
                    generator: `Diplodoc Platform vDIPLODOC-VERSION`,
                });

                const meta = service.get('test.md' as RelativePath);

                expect(
                    uniq(meta.metadata?.map((metaItem: Hash<unknown>) => metaItem.name)).length,
                ).toEqual(meta.metadata?.length);
            });
        });
    });
});
