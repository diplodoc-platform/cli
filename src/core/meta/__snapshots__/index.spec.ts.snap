// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`meta > service > load > should load test file 1`] = `Promise {}`;
